// @ankr.com/multirpc/dist/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configFromEnv = exports.PROD_CONFIG = exports.LOCAL_CONFIG = exports.STAGING_CONFIG = void 0;
exports.STAGING_CONFIG = {
    confirmationBlocks: 15,
    walletPublicUrl: 'https://staging.protocol.ankr.com/',
    walletPrivateUrl: 'https://staging.protocol.ankr.com/',
    workerUrl: 'https://staging.multi-rpc.com/',
    ankrWalletContractAddress: '0xD099e6Be8B30ACc74E3b4e6E80D5e0dB58291e93',
    ankrTokenContractAddress: '0x65BF0bD516aD41622e45bFD22dAD57ffF828333a',
    publicRpcUrl: 'https://staging.multi-rpc.com/{blockchain}',
    publicWsUrl: '',
    privateRpcUrl: 'https://staging.multi-rpc.com/{blockchain}/{user}',
    privateWsUrl: 'wss://staging.multi-rpc.com/{blockchain}/ws/{user}',
};
exports.LOCAL_CONFIG = Object.assign({}, exports.STAGING_CONFIG, {
    walletPublicUrl: 'http://localhost:8000/',
    walletPrivateUrl: 'http://localhost:8000/',
});
exports.PROD_CONFIG = {
    confirmationBlocks: 15,
    walletPublicUrl: 'http://localhost:8000/',
    walletPrivateUrl: 'http://localhost:8000/',
    workerUrl: 'https://next.multi-rpc.com/',
    ankrWalletContractAddress: '0x4432faEe427AE2CA961D5bfEEFC2EeD4e1f1D784',
    ankrTokenContractAddress: '0x8290333cef9e6d528dd5618fb97a76f268f3edd4',
    publicRpcUrl: 'https://rpc.ankr.com/{blockchain}',
    publicWsUrl: '',
    privateRpcUrl: 'https://rpc.ankr.com/{blockchain}/{user}',
    privateWsUrl: 'https://rpc.ankr.com/{blockchain}/ws/{user}',
};
var configFromEnv = function (env) {
    switch (env) {
        case "local":
            return exports.LOCAL_CONFIG;
        case "staging":
            return exports.STAGING_CONFIG;
        case "prod":
            return exports.PROD_CONFIG;
    }
    throw new Error("There is no config for env (" + env + ")");
};
exports.configFromEnv = configFromEnv;
